app/change_set_persisters/hyrax/change_set_persister.rb

```diff
@@ -3,45 +3,107 @@
-        run_hook(:before_delete, change_set: change_set)
-        persister.delete(resource: change_set.resource)
-        run_hook(:after_delete, change_set: change_set)
-      end
+      run_hook(:before_delete, change_set: change_set)
+      persister.delete(resource: change_set.resource)
+      run_hook(:after_delete, change_set: change_set)
     end
 
+    # Calls save on all change_sets in the array passed in
+    # @param [Array] change_sets An array of any change_set created by Hyrax::DynamicChangeSet
+    # @return [Array] an array of the saved resources, such as GenericWork,
+    #  that is subclassed from Valkyrie::Resource
     def save_all(change_sets:)
+      objects = []
       change_sets.map do |change_set|
-        save(change_set: change_set)
+        objects << save(change_set: change_set)
       end
+      objects
     end
 
-    def buffer_into_index
-      metadata_adapter.persister.buffer_into_index do |buffered_adapter|
-        with(metadata_adapter: buffered_adapter) do |buffered_changeset_persister|
-          yield(buffered_changeset_persister)
-        end
+    # Calls delete on all change_sets in the arraypassed in
+    # @param [Array] change_sets An array of any change_set created by Hyrax::DynamicChangeSet
+    # @return [Array] an array of the resources that failed to delete, if empty
+    #  all resources were succesfully deleted
+    def delete_all(change_sets:)
+      change_sets.map do |change_set|
+        delete(change_set: change_set)
       end
     end
 
-    def with(metadata_adapter:)
-      yield self.class.new(metadata_adapter: metadata_adapter, storage_adapter: storage_adapter)
+    # Given a resource in a change_set, recursively deletes all members for the entire
+    # tree of members.  Will delete all members first and then the resource itself
+    # to avoid issues with removing the root of the tree first.  However all other
+    # resources are deleted as found by the query, so issues may arise from the order
+    # the resources were found in, so depending on the complexity of your member tree
+    # you may want to execerise caution in calling this method.
+    # @param [Array] change_sets An array of any change_set created by Hyrax::DynamicChangeSet
+    # @return [Array] an array of the resources that failed to delete, if empty
+    #  all resources were succesfully deleted
+    def delete_self_and_members(change_set:)
+      ids_to_delete = get_ids_of_self_and_all_members(resource: change_set.resource)
+      ids_to_delete.shift # we want to delete members first, self last
+      change_sets = []
+      ids_to_delete.each do |id|
+        resource = metadata_adapter.query_service.find_by(id: id)
+        change_sets << Hyrax::DynamicChangeSet.new(resource)
+      end
+      delete_all(change_sets: change_sets)
+      delete(change_set: change_set)
     end
+
+    private
+
+      def get_ids_of_self_and_all_members(resource:, accumulator: [])
+        accumulator << resource.id
```

Comment:
We're pushing this on just to shift it off on line 86?  That seems unnecessary.

Reply:
This is being with an eye toward moving this over to a query (#1951) where it would be assumed you want the self as well.  We're only shifting this specifically to delay deletion of the original resource to the end of the delete_all.  For non order specific uses (like buffering a resource and all its members somewhere) we'd want the original resource in the array so the function was written from that perspective.